libr/io/p/io_evm.c

```diff
@@ -0,0 +1,686 @@
+	size_t urlmaxlen = sizeof("http://") + sizeof(":65535") + strlen(rioe->host);
+	size_t postfields_len = strlen(tx_trace_req_pattern) + strlen(rioe->tx) + 8;
+	char *postfields = malloc(sizeof(char) * postfields_len);
+
+	if (!postfields) {
+		goto out;
+	}
+
+	if (!rioe->curl) {
+		eprintf("Failed to init curl\n");
+		goto out_free;
+	}
+
+	url = malloc(sizeof(char) * urlmaxlen);
+
+	if (!url) {
+		goto out_free;
+	}
+
+	snprintf(url, urlmaxlen, "http://%s:%d", rioe->host, rioe->port);
+
+	//printf("url is %s\n", url);
+
+	snprintf(postfields, postfields_len,
+				tx_trace_req_pattern, rioe->tx);
+
+	//printf("post body is %s\n", postfields);
+
+	curl_global_init(CURL_GLOBAL_ALL);
+
+	struct resp_data dat = {&rioe->response, 0};
+
+	curl_easy_setopt(rioe->curl, CURLOPT_URL, url);
+	curl_easy_setopt(rioe->curl, CURLOPT_POSTFIELDS, postfields);
+	curl_easy_setopt(rioe->curl, CURLOPT_WRITEFUNCTION, read_response_cb);
+	curl_easy_setopt(rioe->curl, CURLOPT_WRITEDATA, &dat);
+
+	ret = curl_easy_perform(rioe->curl);
+
+	if (ret != CURLE_OK) {
+		eprintf("Failed to get a response from ETH RPC: %s\n",
+				curl_easy_strerror(ret));
+
+		ret = -1;
+	} else {
+		parse_trace(rioe);
+	}
+
+	/*
+	curl_easy_cleanup(rioe->curl);
+	curl_global_cleanup();
+	*/
+
+out_free:
+	free(postfields);
+	free(url);
+out:
+	return ret;
+}
+
+static int evm_read_tx(RIOEvm *rioe) {
+	char *url;
+	int ret = -1;
+	size_t urlmaxlen = sizeof("http://") + sizeof(":65535") + strlen(rioe->host);
+	size_t postfields_len = strlen(get_tx_pattern) + strlen(rioe->tx) + 8;
+	char *postfields = malloc(sizeof(char) * postfields_len);
+
+	if (!postfields) {
+		goto out;
+	}
+
+	if (!rioe->curl) {
+		eprintf("CURL is not initialized\n");
+		goto out_free;
+	}
+
+	url = malloc(sizeof(char) * urlmaxlen);
+
+	if (!url) {
+		goto out_free;
+	}
+
+	snprintf(url, urlmaxlen, "http://%s:%d", rioe->host, rioe->port);
+
+	snprintf(postfields, postfields_len, get_tx_pattern, rioe->tx);
+
+	//printf("get tx post body is %s\n", postfields);
+
+	struct resp_data dat = {&rioe->tx_full, 0};
+	curl_easy_setopt(rioe->curl, CURLOPT_URL, url);
+	curl_easy_setopt(rioe->curl, CURLOPT_POSTFIELDS, postfields);
+	curl_easy_setopt(rioe->curl, CURLOPT_WRITEFUNCTION, read_response_cb);
+	curl_easy_setopt(rioe->curl, CURLOPT_WRITEDATA, &dat);
+
+	ret = curl_easy_perform(rioe->curl);
+
+	if (ret != CURLE_OK) {
+		eprintf("Failed to get a response from ETH RPC: %s\n",
+				curl_easy_strerror(ret));
+
+		ret = -1;
+	} else {
+		//printf("Response is %s\n", rioe->tx_full);
+		parse_transaction(rioe);
+	}
+
+out_free:
+	free(postfields);
+	free(url);
+
+out:
+	return ret;
+}
+
+static int evm_read_code(RIOEvm *rioe) {
+	char *url;
+	int ret = -1;
+    const char *contract_addr;
+	size_t urlmaxlen = sizeof("http://") + sizeof(":65535") + strlen(rioe->host);
+	size_t postfields_len = strlen(code_req_pattern) + strlen(rioe->tx) + 8;
+	char *postfields = malloc(sizeof(char) * postfields_len);
+
+	if (!postfields) {
+		goto out;
```

Comment:
Here it will try to free `url` which wasn't allocated before. Please change this.